src/util/metadata/vo_bit/mod.rs

```diff
@@ -98,45 +98,36 @@ pub fn is_vo_bit_set<VM: VMBinding>(object: ObjectReference) -> bool {
 }
 
 /// Check if an address can be turned directly into an object reference using the VO bit.
-/// If so, return `Some(object)`. Otherwise return `None`.
-pub fn is_vo_bit_set_for_addr<VM: VMBinding>(address: Address) -> Option<ObjectReference> {
-    let potential_object = ObjectReference::from_raw_address(address)?;
-
-    let addr = potential_object.to_address::<VM>();
-
-    // If we haven't mapped VO bit for the address, it cannot be an object
-    if !VO_BIT_SIDE_METADATA_SPEC.is_mapped(addr) {
-        return None;
-    }
-
-    if VO_BIT_SIDE_METADATA_SPEC.load_atomic::<u8>(addr, Ordering::SeqCst) == 1 {
-        Some(potential_object)
-    } else {
-        None
-    }
+pub fn is_vo_bit_set_for_addr<VM: VMBinding>(address: Address) -> bool {
+    _is_vo_bit_set::<true, VM>(address)
 }
 
 /// Check if an address can be turned directly into an object reference using the VO bit.
-/// If so, return `Some(object)`. Otherwise return `None`. The caller needs to ensure the side
-/// metadata for the VO bit for the object is accessed by only one thread.
+/// The caller needs to ensure the side metadata for the VO bit for the object is accessed by only one thread.
 ///
 /// # Safety
 ///
 /// This is unsafe: check the comment on `side_metadata::load`
-pub unsafe fn is_vo_bit_set_unsafe<VM: VMBinding>(address: Address) -> Option<ObjectReference> {
-    let potential_object = ObjectReference::from_raw_address(address)?;
+pub unsafe fn is_vo_bit_set_unsafe<VM: VMBinding>(address: Address) -> bool {
+    _is_vo_bit_set::<false, VM>(address)
+}
 
-    let addr = potential_object.to_address::<VM>();
+fn _is_vo_bit_set<const ATOMIC: bool, VM: VMBinding>(address: Address) -> bool {
+    if let Some(potential_object) = ObjectReference::from_raw_address(address) {
+        let addr = potential_object.to_address::<VM>();
```

Comment:
I see this is a bit complicated.

Would it make things easier if we change the contract of `is_mmtk_object(addr)` so that `addr` must always be 8 byte aligned (or whatever the byte-per-region of `VO_BIT` is)?  In this case, we always check if a VO bit is set at address `addr`.  But when the VM binding does conservative stack scanning, it needs to decide which address to query according to its offset of `ref_to_address`.

I have already thought about `is_mmtk_object` being unsound in the sense that the argument `addr` is not always a valid object reference.  Maybe we just provide a safe variant of `is_vo_bit_set` that checks against the SFT and the memory mapping for the VM binding.  And we explicitly tell the VM that the VO bit is set at `objref.to_address()` rounded down to a multiple of 8 bytes, and let the VM binding figure out how to use it wisely.

Reply:
> Would it make things easier if we change the contract of `is_mmtk_object(addr)` so that `addr` must always be 8 byte aligned (or whatever the byte-per-region of `VO_BIT` is)? In this case, we always check if a VO bit is set at address `addr`. But when the VM binding does conservative stack scanning, it needs to decide which address to query according to its offset of `ref_to_address`.

That will be annoying to the bindings, and expose unnecessary internal details.

I think our current code tries to be unnecessarily general. This results in more bugs, complex code and inefficiency. I would rather introduce some sane assumptions to make both our API and the internal implementation simpler.

Reply:
We already allows the VM binding to set the VO bit in the allocation fast path which inlines the `post_alloc` function.  So the VM binding already knows where VO bit should be set.  So changing the contract to explicitly mention VO bit does not add more burden to the VM binding.

Reply:
> We already allows the VM binding to set the VO bit in the allocation fast path which inlines the `post_alloc` function. So the VM binding already knows where VO bit should be set. So changing the contract to explicitly mention VO bit does not add more burden to the VM binding.

What you said is an optimization. It just copies what we do in mmtk-core, so it naturally assumes our internal implementation. But what we discussed is the API -- it should not expose the internal implementation where possible.